### PR TITLE
BAK-1289 (1) Implement Document Schema

### PR DESCRIPTION
**Pull request by diversemix**
Cautiously starting to implement the new schema. Tested locally and setting up scripts to help migrate and test.

Specification: https://whimsical.com/database-schema-current-6s3aH2TfckVGpGJWhTgJ4j

Marked with green tick on document above:
image

### CODE REVIEW

**Review by THOR300 — APPROVED** [review]
LGTM - assuming the unique string will be what we now call import id?

File: app/db/models/document/physical_document.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+class PhysicalDocument(Base):
+    """A physical document.
+
+    Representation of a document that exists in the real world.
+    Owned and updated only by the pipeline.
+    """
+
+    __tablename__ = "physical_document"
+
+    id = sa.Column(sa.Text, primary_key=True)
```
Comment (joel-wright) [inline app/db/models/document/physical_document.py:14]: I'm not sure why we're setting the primary key to be text at this point; we have nothing sensible to set it to so why not use int?

File: app/db/models/document/physical_document.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    date = sa.Column(sa.DateTime, nullable=False)
+    format = sa.Column(sa.Text, nullable=True)
+
+
+class Language(Base):
+    """Note, moved from deprecated."""
+
+    __tablename__ = "language"
+
+    id = sa.Column(sa.SmallInteger, primary_key=True)
```
Comment (joel-wright) [inline app/db/models/document/physical_document.py:27]: are we trying to save memory? I'm very nervous of ever using small integers...

File: app/db/models/document/physical_document.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    Representation of a document that exists in the real world.
+    Owned and updated only by the pipeline.
+    """
+
+    __tablename__ = "physical_document"
+
+    id = sa.Column(sa.Text, primary_key=True)
+    title = sa.Column(sa.Text, nullable=False)
+    md5_sum = sa.Column(sa.Text, nullable=True)
+    source_url = sa.Column(sa.Text, nullable=True)
```
Comment (joel-wright) [inline app/db/models/document/physical_document.py:17]: I know I set this to `varchar` in the table, but maybe we can use a URL type here?

**Review by joel-wright — COMMENTED** [review]
A few comments about the physical doc field types...

File: app/db/models/document/physical_document.py (hunk 2 above)
Comment (diversemix) [reply app/db/models/document/physical_document.py:27]: ... this was blindly copied from the deprecated module - we can migrate it I suppose?

File: app/db/models/document/physical_document.py (hunk 1 above)
Comment (joel-wright) [reply app/db/models/document/physical_document.py:14]: Is that not pushing law/policy domain information into the application agnostic document schema layer?

File: app/db/models/document/physical_document.py (hunk 2 above)
Comment (joel-wright) [reply app/db/models/document/physical_document.py:27]: ugh, I knew I'd seen this before! Might be worth taking the time to migrate now...

File: app/db/models/document/physical_document.py (hunk 1 above)
Comment (joel-wright) [reply app/db/models/document/physical_document.py:14]: also... we have a single "application schema" at the moment, and while it's true that the family-document <-> physical-document is 1:1 for that app, can we guarantee that constraint holds for all future apps?

File: app/db/models/document/physical_document.py (hunk 3 above)
Comment (joel-wright) [reply app/db/models/document/physical_document.py:17]: ok, this seems to require a custom extension, so scratch that idea

File: app/db/models/document/physical_document.py (hunk 2 above)
Comment (joel-wright) [reply app/db/models/document/physical_document.py:27]: `int2` really scares me

File: app/db/models/document/physical_document.py (hunk 1 above)
Comment (diversemix) [reply app/db/models/document/physical_document.py:14]: 1) It's not really pushing law/policy into this layer - its allowing a upper layer to reference the document schema in a way that it understands - thinking of document id as an abstract identifier that is actually defined in the upper layer... if that makes sense

2) This also re-inforces the above idea - in that even in other apps where the reln is not 1:1 ... the definition of the id needs to really be specified at the upper-scheama level

Comment (diversemix) [reply app/db/models/document/physical_document.py:14]: FTR - conversation concluded to use an ID

File: app/db/models/document/physical_document.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    """A document's languages."""
+
+    __tablename__ = "physical_document_language"
+
+    id = sa.Column(sa.Integer, primary_key=True)
+    language_id = sa.Column(sa.ForeignKey(Language.id), nullable=False)
+    document_id = sa.Column(
+        sa.ForeignKey(PhysicalDocument.id, ondelete="CASCADE"),
+        nullable=False,
+    )
```
Comment (joel-wright) [inline app/db/models/document/physical_document.py:44]: When we spoke earlier I thought we agreed to remove the int id on these join tables...?

```suggestion
    language_id = sa.Column(sa.ForeignKey(Language.id), nullable=False)
    document_id = sa.Column(
        sa.ForeignKey(PhysicalDocument.id, ondelete="CASCADE"),
        nullable=False,
    )
    
    PrimaryKeyConstraint(language_id, document_id)
```

Comment (joel-wright) [reply app/db/models/document/physical_document.py:44]: We can always do this in the follow-up

**Review by joel-wright — APPROVED** [review]

File: app/db/models/document/physical_document.py (hunk 4 above)
Comment (diversemix) [reply app/db/models/document/physical_document.py:44]: FTR - this was done at the point it [was raised](https://github.com/climatepolicyradar/navigator-backend/blob/3855ca5939fa0d5a7741e01ad92c0dc0fb4c7301/app/db/models/document/physical_document.py#L34-L45) ... but closing this off now and merging 1&2 as agreed

**Comment by diversemix** [comment]
Closing as merged (1) and (2)

**Pull request closed without merging**